File: desktop/src/calibration/Calibrator.cpp
#include "calibration/Calibrator.h"

#include <algorithm>
#include <cmath>

namespace singwithme::calibration
{
namespace
{
constexpr float epsilon = 1e-6f;
constexpr float referenceDb = -80.0f;
} // namespace

void Calibrator::start(double sampleRate, float durationSeconds)
{
    sampleRate_ = sampleRate;
    targetDuration_ = durationSeconds;
    processedSamples_ = 0;
    maxAmplitude_ = 0.0f;
}

void Calibrator::processBlock(const float* samples, size_t numSamples)
{
    if (isComplete())
    {
        return;
    }

    for (size_t i = 0; i < numSamples; ++i)
    {
        maxAmplitude_ = std::max(maxAmplitude_, std::abs(samples[i]));
    }

    processedSamples_ += numSamples;
}

bool Calibrator::isComplete() const noexcept
{
    const double totalSamplesNeeded = sampleRate_ * targetDuration_;
    return processedSamples_ >= static_cast<size_t>(totalSamplesNeeded);
}

CalibrationResult Calibrator::result() const noexcept
{
    CalibrationResult res;
    res.isValid = processedSamples_ > 0;
    const float amplitude = std::max(maxAmplitude_, epsilon);
    res.vocalPeakDb = 20.0f * std::log10(amplitude);
    res.noiseFloorDb = referenceDb;
    return res;
}
} // namespace singwithme::calibration

File: web/src/audio/calibrator.ts
const EPSILON = 1e-6;

export interface CalibrationResult {
  noiseFloorDb: number;
  vocalPeakDb: number;
}

export class Calibrator {
  private sampleRate = 48000;
  private targetDuration = 10;
  private processed = 0;
  private maxAmplitude = 0;
  private active = false;

  start(sampleRate: number, durationSeconds = 10) {
    this.sampleRate = sampleRate;
    this.targetDuration = durationSeconds;
    this.processed = 0;
    this.maxAmplitude = 0;
    this.active = true;
  }

  process(frame: Float32Array) {
    if (!this.active) return;
    for (let i = 0; i < frame.length; i += 1) {
      const sample = Math.abs(frame[i]);
      if (sample > this.maxAmplitude) {
        this.maxAmplitude = sample;
      }
    }
    this.processed += frame.length;
    if (this.isComplete()) {
      this.active = false;
    }
  }

  isComplete() {
    return this.processed >= this.sampleRate * this.targetDuration;
  }

  result(): CalibrationResult {
    const amplitude = Math.max(this.maxAmplitude, EPSILON);
    const peakDb = 20 * Math.log10(amplitude);
    return {
      noiseFloorDb: -80,
      vocalPeakDb: peakDb
    };
  }

  store(key = "tunetrix-calibration") {
    const data = this.result();
    localStorage.setItem(key, JSON.stringify({ ...data, timestamp: Date.now() }));
  }

  load(key = "tunetrix-calibration"): CalibrationResult | null {
    const raw = localStorage.getItem(key);
    if (!raw) return null;
    try {
      const parsed = JSON.parse(raw);
      return {
        noiseFloorDb: parsed.noiseFloorDb ?? -80,
        vocalPeakDb: parsed.vocalPeakDb ?? -12
      };
    } catch {
      return null;
    }
  }
}
